client/src/GameState.ts

```diff
@@ -126,6 +205,157 @@ export class GameState {
         this.emitPlayerMove();
     }
 
+    public updateLineClearing() {
+        // check from all four directions, starting from player 0 clockwise
+        if (this.playerId === null) {
+            return; // NOTE this unlikely to happen
+        }
+
+        this.lineCheckSequence.forEach((task) => {
+            const linesToClear = this.scanLinesToClear(task);
+            if (linesToClear.length === 0) {
+                return;
+            }
+            // create a map to record by how much we should fall/offset each lines after removing
+            const linesToFall = this.prepareLinesToFall(task, linesToClear);
+            this.removeLines(task, linesToClear);
+            this.fallLines(task, linesToFall);
+
+            // TODO scoring
+        });
+    }
+
+    private removeLines(task: LineCheckTask, linesToClear: Array<number>) {
+        if (task.axis === LineCheckAxis.Row) {
+            for (const row of task.range) {
+                // scan through each row
+                if (linesToClear.includes(row)) {
```

Comment:
Why iterate over `task.range`? Why not just iterate over `linesToClear`?

Reply:
This function is meant to create a map that tells other uncleared lines how many blocks they should fall, so it's basically a stack iteration, accumulating on the `offset` variable, assigning to each of the remaining lines.